### PR TITLE
Support check committed txns before catalog drop meta, like db, table, partition etc

### PR DESCRIPTION
**Pull request by caiconghui**
This PR is to ensure that dropped db , table or partition can be with normal state after recovered by user.  Commited txns can not be aborted, because the partitions's commited versions have been changed, and some tablets may already have new visible versions. If user just don't want the meta(db, table or partition) anymore, just use dropp instead of drop to skip committed txn check.

### CODE REVIEW

**Comment by caiconghui** [comment]
for #3988

File: fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java [hunk 1]
```diff
@@ -716,7 +716,6 @@ public void processBatchDropRollup(List<AlterClause> dropRollupClauses, Database
 
     public void processDropMaterializedView(DropMaterializedViewStmt dropMaterializedViewStmt, Database db,
             OlapTable olapTable) throws DdlException, MetaNotFoundException {
-        db.writeLock();
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java:719]: Add a `Preconditions.checkState( db.isWriteLockHeldByCurrentThread())` to NOTICE.

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 2]
```diff
@@ -2666,6 +2666,13 @@ public void dropDb(DropDbStmt stmt) throws DdlException {
             Database db = this.fullNameToDb.get(dbName);
             db.writeLock();
             try {
+                if (stmt.isNeedCheckCommitedTxns()) {
+                    if (Catalog.getCurrentCatalog().getGlobalTransactionMgr().existCommittedTxns(db.getId(), null, null)) {
+                        throw new DdlException("There are still some commited txns cannot be aborted in db ["
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/Catalog.java:2671]: There are still some transactions in the COMMITTED state waiting to be completed. The database cannot be dropped. If you want to force drop(cannot be recovered), please use "DROPP database";

File: fe/src/main/java/org/apache/doris/analysis/DropTableStmt.java [hunk 3]
```diff
@@ -61,6 +64,10 @@ public boolean isView() {
         return isView;
     }
 
+    public boolean isNeedCheckCommitedTxns() {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/DropTableStmt.java:67]:
```suggestion
    public boolean isNeedCheckCommittedTxns() {
```

File: fe/src/main/java/org/apache/doris/analysis/DropTableStmt.java [hunk 4]
```diff
@@ -32,17 +32,20 @@
     private boolean ifExists;
     private final TableName tableName;
     private final boolean isView;
+    private boolean needCheckCommitedTxns;
 
-    public DropTableStmt(boolean ifExists, TableName tableName) {
+    public DropTableStmt(boolean ifExists, TableName tableName, boolean needCheckCommitedTxns) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/DropTableStmt.java:37]:
```suggestion
    public DropTableStmt(boolean ifExists, TableName tableName, boolean needCheckCommittedTxns) {
```

File: fe/src/main/java/org/apache/doris/analysis/DropTableStmt.java [hunk 5]
```diff
@@ -32,17 +32,20 @@
     private boolean ifExists;
     private final TableName tableName;
     private final boolean isView;
+    private boolean needCheckCommitedTxns;
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/DropTableStmt.java:35]:
```suggestion
    private boolean needCheckCommittedTxns;
```

File: fe/src/main/java/org/apache/doris/analysis/DropTableStmt.java [hunk 6]
```diff
@@ -32,17 +32,20 @@
...
 
-    public DropTableStmt(boolean ifExists, TableName tableName) {
+    public DropTableStmt(boolean ifExists, TableName tableName, boolean needCheckCommitedTxns) {
         this.ifExists = ifExists;
         this.tableName = tableName;
         this.isView = false;
+        this.needCheckCommitedTxns = needCheckCommitedTxns;
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/analysis/DropTableStmt.java:41]:
```suggestion
        this.needCheckCommittedTxns = needCheckCommittedTxns;
```

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 7]
```diff
@@ -3308,6 +3315,20 @@ public void dropPartition(Database db, OlapTable olapTable, DropPartitionClause
         if (isTempPartition) {
             olapTable.dropTempPartition(partitionName, true);
         } else {
+            if (clause.isNeedCheckCommitedTxns()) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/Catalog.java:3318]: Duplicated `if clause`?

**Comment by morningman** [comment]
If I use `DROPP Table` to forcibly drop a table with COMMITTED txns.
Can those COMMITTED txns be cleaned automatically?

**Comment by caiconghui** [comment]
> If I use `DROPP Table` to forcibly drop a table with COMMITTED txns.
> Can those COMMITTED txns be cleaned automatically?

@morningman  in fe, the txn will be publish timeout, and publish version daemon will try to finish transaction, it find that table have been dropped and will remove table committed info from txn, the txn will be visible. In be, txns seem that it wiil be cleaned by reporter handler?

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java (hunk 7 above)
Comment (caiconghui) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: my bad, I will fix it.

**Comment by chaoyli** [comment]
I found if perform DROP database force, it's committed version will be recovered by RECOVER command?

**Comment by caiconghui** [comment]
> I found if perform DROP database force, it's committed version will be recovered by RECOVER command?

@chaoyli I will fix it, make force drop operation do not recycle meta.

**Review by chaoyli — DISMISSED** [review]

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 8]
```diff
@@ -2666,6 +2666,13 @@ public void dropDb(DropDbStmt stmt) throws DdlException {
             Database db = this.fullNameToDb.get(dbName);
             db.writeLock();
             try {
+                if (stmt.isNeedCheckCommittedTxns()) {
+                    if (Catalog.getCurrentCatalog().getGlobalTransactionMgr().existCommittedTxns(db.getId(), null, null)) {
+                       throw new DdlException("There are still some transactions in the COMMITTED state waiting to be completed. " +
+                               "The database [" + dbName +"] cannot be dropped. If you want to forcibly drop(cannot be recovered)," +
+                               " please use \"DROP database force\".");
+                    }
```
[the remark this answers is not in the document]
Comment (caiconghui) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: ok

File: fe/src/main/cup/sql_parser.cup [hunk 9]
```diff
@@ -879,7 +879,11 @@ alter_table_clause ::=
     :}
     | KW_DROP opt_tmp:isTempPartition KW_PARTITION opt_if_exists:ifExists ident:partitionName
     {:
-        RESULT = new DropPartitionClause(ifExists, partitionName, isTempPartition);
+        RESULT = new DropPartitionClause(ifExists, partitionName, isTempPartition, !isTempPartition);
+    :}
+    | KW_DROPP opt_tmp:isTempPartition KW_PARTITION opt_if_exists:ifExists ident:partitionName
```
[the remark this answers is not in the document]
Comment (caiconghui) [reply fe/src/main/cup/sql_parser.cup]: @chaoyli I has fixed it before. old commit version?

**Review by chaoyli — APPROVED** [review]

**Pull request merged**